### PR TITLE
arrow-cpp: refactor source fetching

### PR DESCRIPTION
**Pull request by cpcloud**
###### Motivation for this change

Follow up from https://github.com/NixOS/nixpkgs/pull/143422#discussion_r739885201

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cpcloud** [comment]
@veprbl Any chance you'd be able to review this? I am working on a couple follow ups that will enable some of the additional features (Flight, S3, and GCS support).

**Review by veprbl — APPROVED** [review]
Diff LGTM

**Comment by veprbl** [comment]
Result of `nixpkgs-review pr 144051` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>arrow-cpp</li>
  </ul>
</details>

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 1]
```diff
@@ -31,22 +31,23 @@ in stdenv.mkDerivation rec {
...
   };
 
-  ARROW_MIMALLOC_URL = fetchurl {
-    # From
-    # ./cpp/cmake_modules/ThirdpartyToolchain.cmake
-    # ./cpp/thirdparty/versions.txt
-    url =
-      "https://github.com/microsoft/mimalloc/archive/v1.7.2.tar.gz";
-    hash = "sha256-sZEuNUVlpLaYQQ91g8D4OTSm27Ot5Uq33csVaTIJNr0=";
+  ARROW_XSIMD_URL = fetchFromGitHub {
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/arrow-cpp/default.nix:46]: Does this gold a path to the nix store or the content itself?

Comment (cpcloud) [reply pkgs/development/libraries/arrow-cpp/default.nix:46]: I assumed it was the nix store. Can derivations point to some other location?

**Comment by veprbl** [comment]
Result of `nixpkgs-review pr 144051` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>21 packages built:</summary>
<br>- arrow-cpp
<br>- python38Packages.datasets
<br>- python38Packages.google-cloud-bigquery
<br>- python38Packages.intake
<br>- python38Packages.intake-parquet
<br>- python38Packages.pyarrow
<br>- python38Packages.snowflake-connector-python
<br>- python38Packages.snowflake-sqlalchemy
<br>- python38Packages.spacy-transformers
<br>- python38Packages.tokenizers
<br>- python38Packages.transformers
<br>- python39Packages.datasets
<br>- python39Packages.google-cloud-bigquery
<br>- python39Packages.intake
<br>- python39Packages.intake-parquet
<br>- python39Packages.pyarrow
<br>- python39Packages.snowflake-connector-python
<br>- python39Packages.snowflake-sqlalchemy
<br>- python39Packages.spacy-transformers
<br>- python39Packages.tokenizers
<br>- python39Packages.transformers
</details>

**Comment by cpcloud** [comment]
@SuperSandro2000 can this be merged?

**Pull request merged**